grpc/runtime/src/main/scala/io/buoyant/grpc/runtime/DecodingStream.scala

```diff
@@ -0,0 +1,288 @@
+
+  def apply[T](
+    rsp: h2.Response,
+    decodeF: ByteBuffer => T
+  ): Stream[T] = new DecodingStream[T] {
+    protected[this] val frames: h2.Stream = rsp.stream
+    protected[this] def decoder: ByteBuffer => T = decodeF
+  }
+
+  private sealed trait RecvState
+
+  private object RecvState {
+
+    case class Buffer(
+      /** The current gRPC message header, if one has been parsed */
+      header: Option[Header] = None,
+      /** Unparsed bytes */
+      buf: Buf = Buf.Empty,
+      /** Tracks how many bytes are consumed and when the underling data may be released */
+      releaser: Releaser = Releaser.Empty
+    ) extends RecvState
+
+    val Empty: RecvState = Buffer()
+  }
+
+  /**
+   * Keeps track of when underlying h2 frames should be released.
+   *
+   * As bytes are read from H2 frames, they are _tracked_. As headers
+   * and messages are read, bytes are _consumed_. When a message is
+   * being constructed, it is paired with a _releasable_ function that
+   * frees the underlying H2 frames as needed.
+   */
+  private trait Releaser {
+
+    /**
+     * Add the given H2 frame to be tracked and released.
+     */
+    def track(frame: h2.Frame): Releaser
+
+    /**
+     * Mark `n` bytes as consumed.
+     */
+    def consume(n: Int): Releaser
+
+    /**
+     * Reap all consumed frames into a releasable function, and return
+     * the updated state afterwards.
+     */
+    def releasable(): (Releaser, Releaser.Func)
+  }
+
+  private object Releaser {
+    type Func = () => Future[Unit]
+    val NopFunc: Func = () => Future.Unit
+
+    class Underflow extends Exception("released too many bytes") with NoStackTrace
+
+    object Empty extends Releaser {
+      override def track(f: h2.Frame): Releaser = mk(f)
+
+      override def consume(n: Int): Releaser =
+        if (n == 0) this
+        else throw new Underflow
+
+      override def releasable(): (Releaser, Func) = (this, NopFunc)
+    }
+
+    private[this] case class FrameReleaser(
+      remaining: Int,
+      release: () => Future[Unit],
+      next: Releaser
+    ) extends Releaser {
+      require(remaining >= 0)
+
+      override def track(f: h2.Frame): Releaser =
+        copy(next = next.track(f))
+
+      override def consume(n: Int): Releaser =
+        if (n == 0) this
+        else if (n >= remaining) {
+          // Leave a 0-length releasable, so that `release` is chained
+          // into the next `releasable` call.
+          copy(
+            remaining = 0,
+            next = next.consume(n - remaining)
+          )
+        } else copy(remaining = remaining - n)
+
+      override def releasable(): (Releaser, Func) =
+        if (remaining == 0) {
+          val (rest, releaseRest) = next.releasable()
+          val doRelease = () => release().before(releaseRest())
+          rest -> doRelease
+        } else {
+          // This frame isn't yet releasable, but we require that this
+          // sub-slice is released before releasing the frame:
```

Comment:
Oh, I get it.  The doRelease func returned here is associated with the previous message.  That message must be released before we can release this frame.  This stuff is hard to get my head around.